File: packages/client-python/src/rfnry_chat_client/handler/stream.py
from __future__ import annotations

import secrets
from datetime import UTC, datetime
from types import TracebackType
from typing import TYPE_CHECKING, Any, Literal

from rfnry_chat_protocol import (
    Identity,
    MessageEvent,
    ReasoningEvent,
    StreamDeltaFrame,
    StreamEndFrame,
    StreamError,
    StreamStartFrame,
    StreamTargetType,
    TextPart,
)

if TYPE_CHECKING:
    from rfnry_chat_client.client import ChatClient


class Stream:
    def __init__(
        self,
        *,
        client: ChatClient,
        thread_id: str,
        run_id: str,
        author: Identity,
        target_type: StreamTargetType,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        self._client = client
        self._thread_id = thread_id
        self._run_id = run_id
        self._author = author
        self._target_type = target_type
        self._metadata = metadata or {}
        self._event_id = f"evt_{secrets.token_hex(8)}"
        self._buffer: list[str] = []
        self._started = False
        self._finalized: MessageEvent | ReasoningEvent | None = None

    @property
    def event_id(self) -> str:
        return self._event_id

    @property
    def finalized_event(self) -> MessageEvent | ReasoningEvent | None:
        return self._finalized

    async def __aenter__(self) -> Stream:
        frame = StreamStartFrame(
            event_id=self._event_id,
            thread_id=self._thread_id,
            run_id=self._run_id,
            target_type=self._target_type,
            author=self._author,
        )
        await self._client.socket.send_stream_start(frame.model_dump(mode="json", by_alias=True))
        self._started = True
        return self

    async def write(self, text: str) -> None:
        if not self._started:
            raise RuntimeError("stream.write called before __aenter__")
        if not text:
            return
        self._buffer.append(text)
        frame = StreamDeltaFrame(
            event_id=self._event_id,
            thread_id=self._thread_id,
            text=text,
        )
        await self._client.socket.send_stream_delta(frame.model_dump(mode="json", by_alias=True))

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> Literal[False]:
        error: StreamError | None = None
        if exc is not None:
            error = StreamError(code="handler_error", message=str(exc))
        end_frame = StreamEndFrame(
            event_id=self._event_id,
            thread_id=self._thread_id,
            error=error,
        )
        await self._client.socket.send_stream_end(end_frame.model_dump(mode="json", by_alias=True))

        if exc is None:
            content = "".join(self._buffer)
            final: MessageEvent | ReasoningEvent
            if self._target_type == "message":
                final = MessageEvent(
                    id=self._event_id,
                    thread_id=self._thread_id,
                    run_id=self._run_id,
                    author=self._author,
                    created_at=datetime.now(UTC),
                    metadata=self._metadata,
                    content=[TextPart(text=content)],
                )
            else:
                final = ReasoningEvent(
                    id=self._event_id,
                    thread_id=self._thread_id,
                    run_id=self._run_id,
                    author=self._author,
                    created_at=datetime.now(UTC),
                    metadata=self._metadata,
                    content=content,
                )
            await self._client.emit_event(final)
            self._finalized = final
        return False

File: packages/client-react/src/stream.ts
import type { Identity, MessageEvent, ReasoningEvent } from '@rfnry/chat-protocol'
import type { SocketTransport } from './transport/socket'

export type StreamOptions = {
  socket: SocketTransport
  threadId: string
  runId: string
  author: Identity
  targetType: 'message' | 'reasoning'
  metadata?: Record<string, unknown>
  onFinalEvent?: (event: MessageEvent | ReasoningEvent) => Promise<void> | void
}

export class ChatStream {
  readonly eventId: string
  private readonly socket: SocketTransport
  private readonly threadId: string
  private readonly runId: string
  private readonly author: Identity
  private readonly targetType: 'message' | 'reasoning'
  private readonly metadata: Record<string, unknown>
  private readonly onFinalEvent?: StreamOptions['onFinalEvent']
  private buffer: string[] = []
  private started = false
  private ended = false

  constructor(opts: StreamOptions) {
    this.socket = opts.socket
    this.threadId = opts.threadId
    this.runId = opts.runId
    this.author = opts.author
    this.targetType = opts.targetType
    this.metadata = opts.metadata ?? {}
    this.onFinalEvent = opts.onFinalEvent
    this.eventId = `evt_${randomHex(8)}`
  }

  async start(): Promise<void> {
    if (this.started) throw new Error('stream already started')
    await this.socket.sendStreamStart({
      eventId: this.eventId,
      threadId: this.threadId,
      runId: this.runId,
      targetType: this.targetType,
      author: authorWire(this.author),
    })
    this.started = true
  }

  async write(text: string): Promise<void> {
    if (!this.started) throw new Error('stream.write called before start')
    if (this.ended) throw new Error('stream.write called after end')
    if (!text) return
    this.buffer.push(text)
    await this.socket.sendStreamDelta({
      eventId: this.eventId,
      threadId: this.threadId,
      text,
    })
  }

  async end(error?: {
    code: string
    message: string
  }): Promise<MessageEvent | ReasoningEvent | null> {
    if (this.ended) return null
    this.ended = true
    await this.socket.sendStreamEnd({
      eventId: this.eventId,
      threadId: this.threadId,
      error: error ?? null,
    })
    if (error) return null

    const content = this.buffer.join('')
    const createdAt = new Date().toISOString()
    let final: MessageEvent | ReasoningEvent
    if (this.targetType === 'message') {
      final = {
        id: this.eventId,
        threadId: this.threadId,
        runId: this.runId,
        author: this.author,
        createdAt,
        metadata: this.metadata,
        recipients: null,
        type: 'message',
        content: [{ type: 'text', text: content }],
      }
    } else {
      final = {
        id: this.eventId,
        threadId: this.threadId,
        runId: this.runId,
        author: this.author,
        createdAt,
        metadata: this.metadata,
        recipients: null,
        type: 'reasoning',
        content,
      }
    }
    if (this.onFinalEvent) {
      await this.onFinalEvent(final)
    }
    return final
  }
}

function randomHex(bytes: number): string {
  const arr = new Uint8Array(bytes)
  crypto.getRandomValues(arr)
  return Array.from(arr, (b) => b.toString(16).padStart(2, '0')).join('')
}

function authorWire(identity: Identity): Record<string, unknown> {
  return {
    role: identity.role,
    id: identity.id,
    name: identity.name,
    metadata: identity.metadata,
  }
}
